tests/src/providers/testqgsvirtualrasterprovider.cpp

```diff
@@ -0,0 +1,232 @@
+void TestQgsVirtualRasterProvider::initTestCase()
+{
+  // init QGIS's paths - true means that all path will be inited from prefix
+  QgsApplication::init();
+  QgsApplication::initQgis();
+
+  mTestDataDir = QStringLiteral( TEST_DATA_DIR ) + '/'; //defined in CmakeLists.txt
+  mReport = QStringLiteral( "<h1>Virtual Raster Provider Tests</h1>\n" );
+
+  QString demFileName = mTestDataDir + "raster/dem.tif";
+  QFileInfo demRasterFileInfo( demFileName );
+  mDemRasterLayer = new QgsRasterLayer( demRasterFileInfo.filePath(),
+                                        demRasterFileInfo.completeBaseName() );
+
+  QString landsatFileName = mTestDataDir + "landsat.tif";
+  QFileInfo landsatRasterFileInfo( landsatFileName );
+  mLandsatRasterLayer = new QgsRasterLayer( landsatRasterFileInfo.filePath(),
+      landsatRasterFileInfo.completeBaseName() );
+}
+
+void TestQgsVirtualRasterProvider::validLayer()
+{
+  QgsRasterLayer::LayerOptions options;
+
+  std::unique_ptr< QgsRasterLayer > layer = std::make_unique< QgsRasterLayer >(
+        mTestDataDir + QStringLiteral( "raster/dem.tif" ),
+        QStringLiteral( "layer" ),
+        QStringLiteral( "virtualraster" ),
+        options
+      );
+
+  QVERIFY( ! layer->isValid() );
+}
+
+//runs after all tests
+void TestQgsVirtualRasterProvider::cleanupTestCase()
+{
+  QgsApplication::exitQgis();
+  QString myReportFile = QDir::tempPath() + "/qgistest.html";
+  QFile myFile( myReportFile );
+  if ( myFile.open( QIODevice::WriteOnly | QIODevice::Append ) )
+  {
+    QTextStream myQTextStream( &myFile );
+    myQTextStream << mReport;
+    myFile.close();
+  }
+}
+
+void TestQgsVirtualRasterProvider::testUriProviderDecoding()
+{
+
+  QgsRasterDataProvider::VirtualRasterParameters decodedParams = QgsVirtualRasterProvider::decodeVirtualRasterProviderUri( QStringLiteral( "?crs=EPSG:4326&extent=18.6662979442000001,45.7767014376000034,18.7035979441999984,45.8117014376000000&width=373&height=350&formula=\"dem@1\" + 200&dem:uri=path/to/file&dem:provider=gdal&landsat:uri=path/to/landsat&landsat:provider=gdal" ) );
+
+  QCOMPARE( decodedParams.width, 373 );
+  QCOMPARE( decodedParams.height, 350 );
+  QCOMPARE( decodedParams.extent, QgsRectangle( 18.6662979442000001, 45.7767014376000034, 18.7035979441999984, 45.8117014376000000 ) );
+  QCOMPARE( decodedParams.crs, QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:4326" ) ) );
+  QCOMPARE( decodedParams.formula, QStringLiteral( "\"dem@1\" + 200" ) );
+
+  QCOMPARE( decodedParams.rInputLayers.at( 1 ).provider, QStringLiteral( "gdal" ) );
+  QCOMPARE( decodedParams.rInputLayers.at( 0 ).provider, QStringLiteral( "gdal" ) );
+
+  qDebug() << QStringLiteral( "Raster layer: name, uri, provider" );
+  for ( int i = 0; i < decodedParams.rInputLayers.size() ; ++i )
+  {
+    qDebug() << decodedParams.rInputLayers.at( i ).name << ", " <<
+             decodedParams.rInputLayers.at( i ).uri  << ", " <<
+             decodedParams.rInputLayers.at( i ).provider;
+  }
```

Comment:
instead of qDebugs, let's add checks for the expected values of `name` and `uri` as well.